### PR TITLE
New Swap Spell - Wizard

### PR DESCRIPTION
**Pull request by Comxy**
## About The Pull Request
This PR will add a new spell called swap. Unlike mindswap, this spell will allow you to change locations with any living being within nine tiles. It can be bought from the spell book for 1 TC since it is worse than jaunt but still has its own unique aspects like being able to time shuttle gibs.

## Why It's Good For The Game
It is cool to have a new spell every once in a while. I was planning to make more but I thought I would keep it at one spell for each PR.
swap0
switch

## Changelog

:cl:
add: Swap Spell - this spell allows you to swap location with any living being.
soundadd: Swap Sounds
imageadd: Swap Icons
/:cl:

### CODE REVIEW

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
+
+/datum/action/cooldown/spell/pointed/swap
+	name = "Swap"
+	desc = "This spell instills a deep terror in your target, temporarily chilling and blinding it."
```
Comment (Time-Green) [inline code/modules/spells/spell_types/pointed/swap.dm:4]: Don’t forgot to update this one

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	button_icon_state = "swap"
+	ranged_mousepointer = 'icons/effects/mouse_pointers/swap_target.dmi'
+
+	school = SCHOOL_EVOCATION
```
Comment (tralezab) [inline code/modules/spells/spell_types/pointed/swap.dm:8]: And this. you're looking for `SCHOOL_TRANSLOCATION`

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	cooldown_time = 30 SECONDS
+	cooldown_reduction_per_rank = 6 SECONDS
```
Comment (tralezab) [inline code/modules/spells/spell_types/pointed/swap.dm:10]: 5 levels of this gives a cooldown time of 0 seconds, have you tested to make sure this won't happen in game?

**Review by tralezab — CHANGES_REQUESTED** [review]

File: code/modules/spells/spell_types/pointed/swap.dm (hunk 3 above)
Comment (Comxy) [reply code/modules/spells/spell_types/pointed/swap.dm:10]: The rank starts at 30, so level five is minus 24 = 6

**Comment by weesnaw-ops** [comment]
This spell is robeless, right?

**Comment by Shadowflame909** [comment]
Sleep: Swapping with someone in a hard to escape space area
Woke: Swapping with the captain as soon as sec tries to stun you/fire at you with multiple lasers/throw a bomb spear at you

This will be a very fun spell to play with in this sandbox. Great idea!

**Comment by Comxy** [comment]
> This spell is robeless, right?

Now you can, considering it isn't that powerful and mostly conditional, people also might want to use it while being shapeshifted. Gives it some extra use.

**Review by MrMelbert — COMMENTED** [review]
Really cool!

**Comment by PowerfulBacon** [comment]
Seems way too situation (only real use is to swap with people if you are in space). Perhaps it should swap status effects and things like stun, so if you get stunned you can swap with someone and put them in a stunned position getting beaten by sec instead of yourself.

**Comment by moocowswag** [comment]
I think its fine as a situational spell, at the end of the day its a cheap extra button you can use as an extra escape tool.

Also it works on any living being right?

Sentient or not

So you could animate stuff or spam bees to give yourself another point that you can teleport to nearby.

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		to_chat(owner, span_warning("The spell had no effect!"))
+		to_chat(cast_on, span_warning("You feel space bending, but it rapidly dissipates."))
+		return FALSE
+
+	to_chat(cast_on, span_userdanger("You feel space bending."))
+	if(ispath(smoke_type, /datum/effect_system/fluid_spread/smoke))
+		var/datum/effect_system/fluid_spread/smoke/smoke = new smoke_type()
+		smoke.set_up(smoke_amt, holder = owner, location = get_turf(owner))
+		smoke.start()
+	var/target_location = cast_on.loc
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:41]:
```suggestion
	var/turf/target_location = cast_on.loc
```
Then we can check if it's a valid turf, too.
And do the same for owner

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+	smoke_type = /datum/effect_system/fluid_spread/smoke
+	smoke_amt = 0
+
+/datum/action/cooldown/spell/pointed/swap/is_valid_target(atom/cast_on)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!isliving(cast_on))
+		to_chat(owner, span_warning("You can only swap locations with living beings!"))
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:25]: would balloon alerts be better?

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 6]
```diff
@@ -0,0 +1,45 @@
+
+/datum/action/cooldown/spell/pointed/swap
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:2]:
```suggestion
/datum/action/cooldown/spell/pointed/swap
```
while im bugging you, might as well

File: code/modules/antagonists/wizard/equipment/spellbook_entries/mobility.dm [hunk 7]
```diff
@@ -30,6 +30,13 @@
 	spell_type = /datum/action/cooldown/spell/jaunt/ethereal_jaunt
 	category = "Mobility"
 
+/datum/spellbook_entry/swap
+	name = "Swap"
+	desc = "Switch places with any living target."
```
Comment (JohnFulpWillard) [inline code/modules/antagonists/wizard/equipment/spellbook_entries/mobility.dm:35]: You should probably mention that they have to be in sight, first thought I had on reading this was 'anyone on the z level'

File: code/modules/spells/spell_types/pointed/swap.dm (hunk 5 above)
Comment (Comxy) [reply code/modules/spells/spell_types/pointed/swap.dm:26]: I think they might get in the way of combat situations and might be annoying with miss-clicks and stuff

**Comment by Comxy** [comment]
Now you can mark a second target to make the spell a bit stronger. You will always teleport to the location your target is at, but the primary target will go to the secondary, and secondary to yourself.

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+		do_teleport(owner, target_location, no_effects = TRUE, channel = TELEPORT_CHANNEL_MAGIC)
+		second_target.playsound_local(get_turf(second_target), 'sound/magic/swap.ogg', 50, 1)
+		cast_on.playsound_local(get_turf(cast_on), 'sound/magic/swap.ogg', 50, 1)
+		owner.playsound_local(get_turf(owner), 'sound/magic/swap.ogg', 50, 1)
+	else
+		do_teleport(cast_on, owner.loc, no_effects = TRUE, channel = TELEPORT_CHANNEL_MAGIC)
+		do_teleport(owner, target_location, no_effects = TRUE, channel = TELEPORT_CHANNEL_MAGIC)
+		cast_on.playsound_local(get_turf(cast_on), 'sound/magic/swap.ogg', 50, 1)
+		owner.playsound_local(get_turf(owner), 'sound/magic/swap.ogg', 50, 1)
+	second_target = NONE
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:82]:
```suggestion
	second_target = null
```

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+			return FALSE
+		if(!isliving(click_target) && isturf(click_target))
+			// Find any living being in the list. We aren't picky, it's aim assist after all
+			click_target = locate(/mob/living) in click_target
+			if(!click_target)
+				to_chat(owner, span_warning("You can only select living beings as secondary target!"))
+				return FALSE
+		if(click_target == owner)
+			to_chat(owner, span_notice("You cancel your secondary swap target!"))
+			second_target = NONE
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:44]:
```suggestion
			second_target = null
```

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+	if(LAZYACCESS(params2list(params), RIGHT_CLICK))
+		if(!IsAvailable())
+			return FALSE
+		if(!target)
+			return FALSE
+		if(!isliving(click_target) && isturf(click_target))
+			// Find any living being in the list. We aren't picky, it's aim assist after all
+			click_target = locate(/mob/living) in click_target
+			if(!click_target)
+				to_chat(owner, span_warning("You can only select living beings as secondary target!"))
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:40]: maybe use balloon alerts here?

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	cast_range = 9
+	invocation = "FRO' BRT'TRO, DA!"
+	invocation_type = INVOCATION_SHOUT
+	spell_requirements = SPELL_REQUIRES_NO_ANTIMAGIC|SPELL_REQUIRES_OFF_CENTCOM
+	active_msg = "You prepare to swap locations with a target..."
+
+	smoke_type = /datum/effect_system/fluid_spread/smoke
+	smoke_amt = 0
+	/// A variable for holding the second selected target with right click.
+	var/mob/living/second_target
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:19]: would this be better to use weakref instead?

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		smoke.set_up(smoke_amt, holder = owner, location = get_turf(owner))
+		smoke.start()
+	var/turf/target_location = cast_on.loc
+	if(!isnull(second_target) && get_dist(owner, second_target) <= cast_range && !(cast_on == second_target))
+		to_chat(second_target, span_userdanger("You feel space bending."))
+		if(ispath(smoke_type, /datum/effect_system/fluid_spread/smoke))
+			var/datum/effect_system/fluid_spread/smoke/smoke = new smoke_type()
+			smoke.set_up(smoke_amt, holder = owner, location = get_turf(second_target))
+			smoke.start()
+		var/turf/second_location = second_target.loc
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:70]:
```suggestion
		var/turf/second_location = get_turf(second_target)
```

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 13]
```diff
@@ -0,0 +1,82 @@
+/datum/action/cooldown/spell/pointed/swap
+	name = "Swap"
+	desc = "This spell allows you to swap locations with any living being."
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:3]: can you add ``RMB: [rmb explanation]`` here? Like modsuits do currently.

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+		return FALSE
+	return TRUE
+
+/datum/action/cooldown/spell/pointed/swap/InterceptClickOn(mob/living/caller, params, atom/click_target)
+	if(LAZYACCESS(params2list(params), RIGHT_CLICK))
+		if(!IsAvailable())
+			return FALSE
+		if(!target)
+			return FALSE
+		if(!isliving(click_target) && isturf(click_target))
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:36]: What if you click on the user directly, and not the turf? It won't set click_target

File: code/modules/spells/spell_types/pointed/swap.dm [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+			// Find any living being in the list. We aren't picky, it's aim assist after all
+			click_target = locate(/mob/living) in click_target
+			if(!click_target)
+				to_chat(owner, span_warning("You can only select living beings as secondary target!"))
+				return FALSE
+		if(click_target == owner)
```
Comment (JohnFulpWillard) [inline code/modules/spells/spell_types/pointed/swap.dm:42]: would this be better as RMB while you have a target, instead?

**Review by JohnFulpWillard — COMMENTED** [review]
forgot this too

File: code/modules/spells/spell_types/pointed/swap.dm (hunk 10 above)
Comment (Comxy) [reply code/modules/spells/spell_types/pointed/swap.dm:40]: Balloon alerts will probably get annoying so I'll leave them out

File: code/modules/spells/spell_types/pointed/swap.dm (hunk 11 above)
Comment (Comxy) [reply code/modules/spells/spell_types/pointed/swap.dm:19]: I cannot play sounds to a weakref so I'll use this container instead

File: code/modules/spells/spell_types/pointed/swap.dm (hunk 15 above)
Comment (Comxy) [reply code/modules/spells/spell_types/pointed/swap.dm:42]: If someone is moving they might want to click more than once to make sure they have marked the target, that is why to unmark it you will have to click yourself.

File: code/modules/spells/spell_types/pointed/swap.dm (hunk 14 above)
Comment (Comxy) [reply code/modules/spells/spell_types/pointed/swap.dm:36]: It will discard the secondary target

**Review by tralezab — APPROVED** [review]
ogey

**Pull request merged**